---
title: "An Introduction to BSDE"
author: "Mengqi Zhang, Richard Guo"
date: "`r Sys.Date()`"
output:
  pdf_document
vignette: >
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo=TRUE, cache = FALSE, warning = FALSE, 
  message = FALSE, cache.lazy = FALSE
)
library("BSDE")
```

This markdown includes two parts. 

The first parts gives a guide for the core usage of BSDE. 

The second part includes an example of using BSDE on real data. 

# Installation

BSDE can be installed from GitHub.
```{r install,eval=FALSE,echo=TRUE}
# install.packages("devtools")
devtools::install_github("mqzhanglab/BSDE")
```

# Part I: Core function of BSDE

The core function from BSDE is `cal_w2_pval`. It calculates the p-values gene by gene. 
In most situation, the log-transformed expressions are recommended.

`cal_w2_pval` has three required inputs:
`count_per_gene`,  `meta_individual` and `meta_phenotype`.

This function return a list, which contains:
* `pval`: p-value
* `case_bc_ob`: density vector of the Barycenter distribution from cases
* `ctrl_bc_ob`: density vector of the Barycenter distribution from controls

## Example

```{r pure_example_package,eval=TRUE,echo=TRUE}
library("BSDE")
library("ggplot2")
```

Suppose we have 12 subjects (6 cases and 6 controls), where each subject has 10 cells.

```{r pure_example_setting,eval=TRUE,echo=TRUE}
count_per_gene=c(rpois(60,6),c(rpois(60,4)))
meta_individual=paste0("ind",rep(1:12,each=10))
meta_phenotype=factor(c(rep(1,60),rep(0,60)))

# show dataset
table(meta_individual)
table(meta_phenotype)

# expressions
df=data.frame(count=count_per_gene,ind=meta_individual, diagnosis=meta_phenotype)
ggplot(df,aes(x=count,color=diagnosis,fill=diagnosis,group=ind))+ geom_density(alpha=0.05)+ theme_classic()
```

Let us calculate the p-value.
```{r pure_example_res,eval=TRUE,echo=TRUE}
results <- cal_w2_pval(count_per_gene,meta_individual,meta_phenotype)
print(results[[1]])
plot(results[[2]], type="l", col="red", xlab="expression", ylab="density")
lines(results[[3]], col="blue")
legend("topright", c("case", "control"), lwd=c(1,1), col=c("red", "blue"))
```

# Part II: Data analysis example

Here we present a demo simulation here. 
The simulation is based on a real autism dataset from this [paper](https://doi.org/10.1126/science.aav8130).

We generate 30 genes (from a particular cell type) for 20 case subjects and 20 control subjects. 
Each subject comes with 20 cells. 

We simulate 4 different types of differential expressions (DEs) by adding on particular ratios to the basic parameters to get expected fold changes, the rest genes are simulated without size factors and as controls

(1). mean DE: 3 genes are DE in mean but not in var (size factor 1.2).
(2).var DE: 3 genes are DE in var but not mean(size factor 1.2).
(3).multimodality DE: 3 genes are DE due to Multimodality (size factor 0.6).
(4).disp DE: 3 genes are DE in dispersion (size factor 0.2).

Please see our [simulation guide](https://github.com/mqzhanglab/BSDE/tree/main/simulation) for more details. 
```{r meta_simulation,eval=TRUE,echo=TRUE}
sim_matrix=readRDS("../data/simulated/sim_matrix_1.2_1.2_0.6_0.2.rds")
sim_matrix_bulk=readRDS("../data/simulated/sim_matrix_bulk_1.2_1.2_0.6_0.2.rds")
meta=readRDS("../data/simulated/sim_meta_1.2_1.2_0.6_0.2.rds")
```


## Differential Expressed Analysis
We calculate the p-values of the simulated genes through the BSDE method as well as the [MAST](https://www.bioconductor.org/packages/release/bioc/html/MAST.html) (with a widely used protocol mixed-effect model "glmer" & Likelihood-ratio Test) and the [DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).

## BSDE
We also calculate the p-values using the BSDE as follows.We can either run it one-by-one with the major command for BSDE is cal_w2_pval. Or we can start from SingleCellExperiment objects.

```{r BSDE0,eval=FALSE,echo=FALSE}
library("BSDE")

sim_matrix_log = log2(1 + sim_matrix) #log transformed data

dim(sim_matrix_log)
sim_matrix_log[1:2, 1:5]

op_pval = matrix(ncol = 1, nrow = nrow(sim_matrix_log))

print(date())
print(gc())

for (i_g in 1:nrow(sim_matrix_log)) {
  cur_sim = sim_matrix_log[i_g, ]
  cur_ind = meta$individual
  cur_pheno = meta$phenotype
  
  op_pval[i_g] = tryCatch({
    cal_w2_pval(
      count_per_gene = cur_sim,
      meta_individual = cur_ind,
      meta_phenotype = cur_pheno,
      perm_num = 500,
      unif_round_unit = 0.5
    )[[1]]
  }, error = function(e) {
    NA
  })
}
print(date())
print(gc())

rownames(op_pval) = gene_id
length(pval)
pval[1:5]
```

```{r BSDE1,eval=TRUE,echo=TRUE}
library("BSDE")
library("Barycenter")
library("scater")
library("SingleCellExperiment")
library("reticulate")
library("doRNG")
library("doParallel")
py_run_file("../inst/op_functions.py")
```

The input data could be a SingleCellExperiment objects with the "counts" in assays, a rowData named "gene_id" as gene_names and two columns named as "individual" and "condition". The "individual" is a factor that represents individual labels. The "condition" is a numerical that is indicator (contains 0,1).
```{r BSDE2,eval=TRUE,echo=TRUE}
gene_id = rownames(sim_matrix)   #get the gene id from the data
meta$condition <- meta$phenotype
sce=SingleCellExperiment::SingleCellExperiment(list(counts=sim_matrix),colData=meta,rowData=gene_id)
```

Then we use the function BSDE to calculate the p-values.
```{r BSDE2,eval=TRUE,echo=TRUE}
pval=BSDE::BSDE(sce)
```

The result is a list of p-values. To keep the format in accordance with the input, the model failure results are represented as NA.
```{r BSDE2,eval=TRUE,echo=TRUE}
length(pval)
pval[1:5]
```



## DESeq2

For DESeq2 analysis, we construct the bulk RNAseq situation though suming up the counts per gene, per cluster, per individual, per 1000 cells.

```{r deseq2,eval=FALSE}
library("DESeq2")
cur_info = meta[, c("individual", "phenotype")]
cur_info = unique(cur_info)
rownames(cur_info) = cur_info$individual
cur_info$phenotype = as.factor(cur_info$phenotype)

# object construction
dds = DESeqDataSetFromMatrix(countData = sim_matrix_bulk,
                             colData = cur_info,
                             design = ~ phenotype)

# observed pvalue calculation
dds = DESeq(dds)
deseq_pval = results(dds)$pvalue
```

## MAST

We take the method of current usage of MAST, we get the p-values based on the Hurdle model ("H" model)
```{r MAST,eval=FALSE}
library("MAST")
library("lme4")

sim_matrix_log = log2(1 + sim_matrix) #log transformed data

dim(sim_matrix_log)
sim_matrix_log[1:2, 1:5]

cell_id = colnames(sim_matrix_log)   #get the cell id from the data
gene_id = rownames(sim_matrix_log)   #get the gene id from the data

fData = data.frame(primerid = gene_id)
cData = data.frame(wellKey = cell_id)

diagnosis = as.character(meta$phenotype) #

diagnosis2=matrix("Control",ncol=1,nrow=length(diagnosis))
diagnosis2[which(diagnosis == "1")] = "Case"
diagnosis= as.factor(diagnosis2)

sca = MAST::FromMatrix(sim_matrix_log, cData, fData)
colData(sca)$cngeneson = as.numeric(meta$CDR)
colData(sca)$diagnosis = as.factor(diagnosis)
colData(sca)$ind = as.factor(meta$individual)

colData(sca)

print(date())
print(gc())
b1 = tryCatch(MAST::zlm(formula = ~ diagnosis + ( 1 | ind ), sca = sca, method = 'glmer', ebayes = FALSE, parallel = TRUE), error = function(e) {NA} )
print(date())
print(gc())

lrt1 = tryCatch(MAST::lrTest(b1, "diagnosis"), error = function(e) {NA} )
MAST_pval1 = tryCatch(apply(lrt1, 1, function(x){x[3,3]}), error = function(e) {NA} )
print(date())
print(gc())


```

## Further Analysis and Method Evaluation
Once we have the p-values, we can do further analysis on the method evaluation.For example, we can calculate the proportion of p-values < 0.05. We can also draw similar plot for power, where power is calculated as the proportion of p-values < 0.05 in permuted data. 

# Session Information
```{r ,eval=TRUE}
sessionInfo()
```

# Reference


Finak, G., McDavid, A., Yajima, M., Deng, J., Gersuk, V., Shalek, A. K., ... & Gottardo, R. (2015). MAST: a flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell RNA sequencing data. Genome biology, 16(1), 1-13.

Love, M. I., Huber, W., & Anders, S. (2014). Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome biology, 15(12), 1-21.

Velmeshev, D., Schirmer, L., Jung, D., Haeussler, M., Perez, Y., Mayer, S., ... & Kriegstein, A. R. (2019). Single-cell genomics identifies cell typeâ€“specific molecular changes in autism. Science, 364(6441), 685-689.
